### PR TITLE
Implement `envtool tests shard` command

### LINKED ISSUE #2692: Implement `envtool tests shard` command
### What should be done?

We want to make CI faster by splitting integration tests into multiple parallel GitHub Actions jobs. Currently, we split them by handler (PostgreSQL, SQLite, etc), but we could also split tests of one handler. For example, we could run `go test -run=^(TestOne|BenchmarkTwo)$` in one job, `go test -run=^(ExampleThree|FuzzFour)$` in another, etc. But we don't want to update our CI configurations every time we add or remove a top-level integration test.

Let's implement `envtool tests shard` command. It would output the value of the `-run` flag that we would then pass to `go test`. For example:

```
$ envtool tests shard -index=0 -total=2

^(TestOne|BenchmarkTwo)$

$ envtool tests shard -index=1 -total=2

^(ExampleThree|FuzzFour)$
```

It then could be used as `go test -run=$(envtool tests shard -index=0 -total=2)`.

A list of test functions could be received using `go test -list=.` command. `envtool tests shard` should call it in the `integartion` directory, split output into lines, sort them, then iterate over them and split them into multiple shards.

This task is only about implementing that command and testing it manually. CI integration is not in the scope.

### Where?

- https://github.com/FerretDB/FerretDB/tree/main/cmd/envtool

### Definition of Done

- `envtool tests shard -index=0 -total=2` command works;
- there are some basic tests that actually run `go test -list=.` and shard results in place.

### PR DESCRIPTION
**Pull request by kropidlowsky**
# Description

Closes #2692.

Implemented `envtool tests shard` command.

- Command example: `go run ./cmd/envtool tests shard --index=0 --total=2`
- Usage of the command with **go test** - **it does not contain required flags for integration tests**: `go test -run regexp="$(go run ../cmd/envtool tests shard --index=0 --total=2)"` - works in the integration folder

## Readiness checklist

- [x] I added/updated unit tests.
- [ ] I added/updated integration/compatibility tests.
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: cmd/envtool/tests.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+// shardTests shards gotten test names from the specified path.
+func shardTests(index, total uint, path string) ([]string, error) {
+	testNames, err := getAllTestNames(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return testNames[index:total], nil
```
Comment (chilagrow) [inline cmd/envtool/tests.go:51]: If index is `2` and total is `3`, it should return 3 items, currently it uses `[2:3]`, and returns 1 item. Let's fix that.

```
$ go run ./cmd/envtool tests shard --index=2 --total=3
^(TestAggregateCompatCollStats)
```
Also, if `total` or `index` is larger than the `testNames` slice, it panics. We should check that.

File: cmd/envtool/tests.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+	cmd := exec.Command("go", "test", "-list=.")
+
+	outputBytes, err := cmd.Output()
+	if err != nil {
+		return tests, err
+	}
+	output := string(outputBytes)
+
+	tests = strings.FieldsFunc(output, func(s rune) bool {
```
Comment (chilagrow) [inline cmd/envtool/tests.go:74]: `strings.Split(...)` is a simpler function of this :hugs:

File: cmd/envtool/tests.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	return testNames[index:total], nil
+}
+
+// getAllTestNames gets all test names from the specified path.
+func getAllTestNames(path string) ([]string, error) {
+	var tests []string
+
+	if err := chdirToRoot(); err != nil {
```
Comment (chilagrow) [inline cmd/envtool/tests.go:58]: Why do we have to change directory to root? Shouldn't `tests shard` be called from top level directory? :thinking:

File: cmd/envtool/tests_test.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	t.Parallel()
+
+	var w bytes.Buffer
+	err := shardIntegrationTests(&w, 0, 2)
+	assert.NoError(t, err)
+	assert.NotNil(t, w)
+	assert.Regexp(t, "^\\^.*\\$$", w.String())
+}
+
+func TestShardTests(t *testing.T) {
```
Comment (chilagrow) [inline cmd/envtool/tests_test.go:35]: Running `task tests-unit` fails on new tests.
```
--- FAIL: TestShardTests (0.00s)
    tests_test.go:39: 
        	Error Trace:	/home/runner/work/FerretDB/FerretDB/cmd/envtool/tests_test.go:39
        	Error:      	Received unexpected error:
        	            	chdir integration: no such file or directory
        	Test:       	TestShardTests
    tests_test.go:40: 
        	Error Trace:	/home/runner/work/FerretDB/FerretDB/cmd/envtool/tests_test.go:40
        	Error:      	"[]" should have 2 item(s), but has 0
        	Test:       	TestShardTests
--- FAIL: TestChdirToRoot (0.00s)
    tests_test.go:61: 
        	Error Trace:	/home/runner/work/FerretDB/FerretDB/cmd/envtool/tests_test.go:61
        	Error:      	Should not be: "/home/runner/work/FerretDB/FerretDB/integration"
        	Test:       	TestChdirToRoot
Debug: --- FAIL: TestGetAllTestNames (0.00s)
    tests_test.go:[47](https://github.com/FerretDB/FerretDB/actions/runs/5149995663/jobs/9273652428?pr=2773#step:6:48): 
        	Error Trace:	/home/runner/work/FerretDB/FerretDB/cmd/envtool/tests_test.go:47
        	Error:      	Received unexpected error:
        	            	chdir integration: no such file or directory
        	Test:       	TestGetAllTestNames
    tests_test.go:[48](https://github.com/FerretDB/FerretDB/actions/runs/5149995663/jobs/9273652428?pr=2773#step:6:49): 
        	Error Trace:	/home/runner/work/FerretDB/FerretDB/cmd/envtool/tests_test.go:48
        	Error:      	Should NOT be empty, but was []
        	Test:       	TestGetAllTestNames
FAIL
```

File: cmd/envtool/tests.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		return tests, err
+	}
+	output := string(outputBytes)
+
+	tests = strings.FieldsFunc(output, func(s rune) bool {
+		return s == '\n'
+	})
+	sort.Strings(tests)
+
+	return tests, nil
```
Comment (chilagrow) [inline cmd/envtool/tests.go:79]: The last item is not a test name
```
|ok       github.com/FerretDB/FerretDB/integration        0.026s)
```
See
```
$ go test -list=.
TestAggregateCommandCompat
TestAggregateCompatStages
...
BenchmarkInsertMany
ok      github.com/FerretDB/FerretDB/integration        0.023s
```

File: cmd/envtool/envtool.go [hunk 6]
```diff
@@ -401,6 +401,12 @@ var cli struct {
 		} `cmd:"" help:"read files"`
 	} `cmd:""`
 	PackageVersion struct{} `cmd:"" help:"Print package version"`
+	Tests          struct {
+		Shard struct {
+			Index uint `help:"Iteration's index" required:""`
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:406]: That's shard index

File: cmd/envtool/envtool.go [hunk 7]
```diff
@@ -401,6 +401,12 @@ var cli struct {
...
+			Total uint `help:"Total number of integration tests to be listed" required:""`
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:407]: That should be the total number of shards. For example, `-index=2 -total=5` means there are 5 shards with an index from 0 to 4, and we want to run a shard with an index 2.

We also should check somewhere that `index < total`

File: cmd/envtool/tests.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+// getAllTestNames gets all test names from the specified path.
+func getAllTestNames(path string) ([]string, error) {
+	var tests []string
+
+	if err := chdirToRoot(); err != nil {
+		return tests, err
+	}
+
+	if err := os.Chdir(path); err != nil {
```
Comment (AlekSi) [inline cmd/envtool/tests.go:62]: Instead of changing the global current directory, it will be much better to set `cmd.Dir`. That should make everything easier.

File: cmd/envtool/tests_test.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	t.Parallel()
+
+	testNames, err := shardTests(0, 2, "integration")
```
Comment (chilagrow) [inline cmd/envtool/tests_test.go:38]: Thanks for adding test! Great start!
We should also test for `(1, 2, "integration")` index that is not zero as well :hugs: 

Also another good tests are testing for large index `(1000, 10, "integration")` and large total `(10, 1000, "integration")`

File: cmd/envtool/tests.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	tests = strings.FieldsFunc(output, func(s rune) bool {
+		return s == '\n'
+	})
+	sort.Strings(tests)
+
+	return tests, nil
+}
+
+// chdirToRoot changes the directory to the root of the repository.
+func chdirToRoot() error {
```
Comment (chilagrow) [inline cmd/envtool/tests.go:83]: When I run suggested command, I receive following directory not found :thinking: 
I'm not sure why we need this function.
```
chi@alma:~/go/src/github.com/chilagrow/FerretDB$ go test -run regexp="$(go run ../cmd/envtool tests shard --index=0 --total=2)"
stat /home/chi/go/src/github.com/chilagrow/cmd/envtool: directory not found
no Go files in /home/chi/go/src/github.com/chilagrow/FerretDB
$ echo $PWD
/home/chi/go/src/github.com/chilagrow/FerretDB
```

File: cmd/envtool/tests.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+// shardIntegrationTests shards integration test names from the specified path.
+func shardIntegrationTests(w io.Writer, index, total uint) error {
+	var output strings.Builder
+
+	testNames, err := shardTests(index, total, "integration")
+	if err != nil {
+		return err
+	}
+	shardedTestNames := testNames[index:total]
```
Comment (chilagrow) [inline cmd/envtool/tests.go:33]: In `shardTests` L51, it takes slice `testNames[index:total]` I think we shouldn't do this here again right?

**Review by chilagrow — COMMENTED** [review]
Thanks for the contribution!

I have few comments. Let` me know if you need any help with this task :hugs:

File: cmd/envtool/tests.go (hunk 10 above)
Comment (kropidlowsky) [reply cmd/envtool/tests.go]: The command works when the current working directory is FerretDB/integration

Comment (kropidlowsky) [reply cmd/envtool/tests.go]: But it is a "shorter" version of the command because it does not contain required flags for intgration tests (e.g. -target-backend)

**Comment by kropidlowsky** [comment]
Hi @chilagrow and @AlekSi,

Could you please take look again?
I have tried to align the code to your reviews + made some small refactoring.

Thanks in advance!

File: cmd/envtool/tests.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+		return nil, fmt.Errorf("Cannot shard when Total is greater than amount of tests (%d > %d)", total, testsLen)
+	}
+
+	sharder := testsLen / total
+	start := sharder * index
+	end := sharder*index + sharder
+
+	if index == total-1 {
+		modulo := testsLen % total
+		end += modulo
```
Comment (chilagrow) [inline cmd/envtool/tests.go:67]: Got it, the remainder will be in the last shard so the last shard could be bigger by at most `total-1` :+1:

File: cmd/envtool/tests_test.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+	assert.NoError(t, err)
+	assert.NotEmpty(t, testNames)
+}
+
+func TestPrepareTestNamesOutput(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Fails when the status is not OK", func(t *testing.T) {
+		output := `TestAggregateCommandCompat
+fail      github.com/FerretDB/FerretDB/integration        0.023s`
```
Comment (chilagrow) [inline cmd/envtool/tests_test.go:137]: Great test :+1:

**Review by chilagrow — COMMENTED** [review]
Looks great. I have minor comments about linter warnings :hugs:

File: cmd/envtool/envtool.go [hunk 14]
```diff
@@ -401,6 +401,12 @@ var cli struct {
 		} `cmd:"" help:"read files"`
 	} `cmd:""`
 	PackageVersion struct{} `cmd:"" help:"Print package version"`
+	Tests          struct {
+		Shard struct {
+			Index uint `help:"Shard index" required:""`
```
[the remark this answers is not in the document]
Comment (kropidlowsky) [reply cmd/envtool/envtool.go]: I am wondering why linter was raising:
```
cmd/envtool/envtool.go:406:15: tag is not aligned, should be: help:"Shard index"            required:"" (tagalign)
			Index uint `help:"Shard index" required:""`
```

When I cut `Tests Shard` Index's help it is fine

[the remark this answers is not in the document]
Comment (chilagrow) [reply cmd/envtool/envtool.go]: I think `tagalign` linter is telling that it wants `required` tag of `Index` and `Total` to be aligned. Let's do that :hugs:
```
			Index uint `help:"Shard index"            required:""`
			Total uint `help:"Total number of shards" required:""`
```

**Review by chilagrow — APPROVED** [review]
Great! Being able to run subsets of tests like this is amazing :hugs: 

```
$ GOFLAGS="-run=$(go run ./cmd/envtool tests shard --index=0 --total=30) -parallel=1" task test-integration-pg
task: [test-integration-pg] go test -count=1 -timeout=35m -race=true -tags=ferretdb_debug,ferretdb_tigris,ferretdb_hana -shuffle=on -coverpkg=../... -coverprofile=integration-pg.txt . -target-backend=ferretdb-pg -target-tls -postgresql-url=postgres://username@127.0.0.1:5432/ferretdb -compat-url='mongodb://username:password@127.0.0.1:47018/?tls=true&tlsCertificateKeyFile=../build/certs/client.pem&tlsCaFile=../build/certs/rootCA-cert.pem'

ok      github.com/FerretDB/FerretDB/integration        11.738s coverage: 40.0% of statements in ../...
```

**Review by AlekSi — APPROVED** [review]
